### PR TITLE
improved duplicative case & jurisdiction handling

### PR DESCRIPTION
**Pull request by ChefAndy**
- capdb/migrations/0005_casemetadata_duplicative.py
    -- added duplicative bool to case metadata model
- capdb/migrations/0006_auto_20170822_1548.py
    -- added duplicative option for citation type
- capdb/models.py
    -- improves jurisdiction handling
    -- translates bad jurisdiction values to good ones in metadata
    -- now handles duplicative regional cases
    -- updated fields as specified above
- fabfile.py
    -- now populates jurisdictions before ingesting metadata
- scripts/helpers.py
    -- contains jurisdiction translations
    -- contains duplicative entry in nsmap
- scripts/ingest_files.py
    -- updates metadata upon ingesting each case
- scripts/ingest_tt_data.py
    -- populates jurisdiction table
- scripts/process_metadata.py
    -- added support for duplicative regional cases

### CODE REVIEW

**Comment by ChefAndy** [comment]
I'll work on getting these tests working.

File: capstone/capdb/models.py [hunk 1]
```diff
@@ -259,56 +260,65 @@ def __str__(self):
 
     def update_case_metadata(self):
         data = get_case_metadata(self.orig_xml)
-        citation, created = Citation.objects.get_or_create(
-            cite=data["citation"],
-            type=data["citation_type"])
+        volume_metadata = VolumeMetadata.objects.get(barcode=data["volume_barcode"])
+        reporter = Reporter.objects.get(id=volume_metadata.reporter_id)
```
Comment (jcushman) [inline capstone/capdb/models.py]: Here it's a little clearer to use model attributes (which should run similar queries behind the scenes):

```
volume_metadata = self.volume.volume_metadata
reporter = volume_metadata.reporter
```

File: capstone/capdb/models.py [hunk 2]
```diff
@@ -259,56 +260,65 @@ def __str__(self):
...
+            case_metadata.decision_date_original = data["decision_date_original"]
+            case_metadata.decision_date = data["decision_date"]
+            case_metadata.docket_number = data["docket_number"]
+
+            if data['volume_barcode'] in special_jurisdiction_cases:
+                case_metadata.jurisdiction = Jurisdiction.objects.get(name=special_jurisdiction_cases[data["volume_barcode"]])
+            else:
+                case_metadata.jurisdiction = Jurisdiction.objects.get(name=jurisdiction_tranlsation[data["jurisdiction"]])
+            
+            print(case_metadata.jurisdiction)
```
Comment (jcushman) [inline capstone/capdb/models.py]: Probably not useful to keep this debugging line in?

File: capstone/capdb/models.py [hunk 3]
```diff
@@ -259,56 +260,65 @@ def __str__(self):
...
+                citation=citation)
 
+            case_metadata.decision_date_original = data["decision_date_original"]
+            case_metadata.decision_date = data["decision_date"]
+            case_metadata.docket_number = data["docket_number"]
+
+            if data['volume_barcode'] in special_jurisdiction_cases:
+                case_metadata.jurisdiction = Jurisdiction.objects.get(name=special_jurisdiction_cases[data["volume_barcode"]])
+            else:
+                case_metadata.jurisdiction = Jurisdiction.objects.get(name=jurisdiction_tranlsation[data["jurisdiction"]])
```
Comment (jcushman) [inline capstone/capdb/models.py]: Typo in `jurisdiction_tranlsation`

File: capstone/capdb/models.py [hunk 4]
```diff
@@ -259,56 +260,65 @@ def __str__(self):
...
+                case_metadata.court = court
+
+            if case_metadata.court and case_metadata.jurisdiction:
+                court.jurisdiction = case_metadata.jurisdiction
+                court.save()
+
+
+        else:
+            citation, created = Citation.objects.get_or_create(
+                cite="{}_{}_{}".format(reporter.short_name, volume_metadata.volume_number, data["volume_barcode"]),
```
Comment (jcushman) [inline capstone/capdb/models.py]: I think these want to be spaces and not underscores? `"{} {} {}"`

File: capstone/scripts/helpers.py [hunk 5]
```diff
@@ -15,6 +16,94 @@
     'info': 'info:lc/xmlns/premis-v2',
 }
 
+jurisdiction_tranlsation= {
```
Comment (jcushman) [inline capstone/scripts/helpers.py]: Typo

File: capstone/scripts/ingest_tt_data.py [hunk 6]
```diff
@@ -91,5 +92,20 @@ def copyModel(source, destination, field_map, dupcheck, dupe_field='id'):
...
+            if source.__name__ == 'Reporters':
+                destination_record.jurisdictions.add(Jurisdiction.objects.get(name=source_record.state))
+                destination_record.save()
+
         except IntegrityError as e:
             print(e)
+
+def populate_jurisdiction():
+    """This populates the jurisdiction table based on what's in the tracking tool stub"""
+    reporters = Reporters.objects.order_by().values('state').distinct()
```
Comment (jcushman) [inline capstone/scripts/ingest_tt_data.py]: Blank `order_by()` can be omitted

File: capstone/scripts/ingest_tt_data.py [hunk 7]
```diff
@@ -91,5 +92,20 @@ def copyModel(source, destination, field_map, dupcheck, dupe_field='id'):
...
+                destination_record.save()
+
         except IntegrityError as e:
             print(e)
+
+def populate_jurisdiction():
+    """This populates the jurisdiction table based on what's in the tracking tool stub"""
+    reporters = Reporters.objects.order_by().values('state').distinct() 
+    for jurisdiction in reporters:
+        if Jurisdiction.objects.filter(name=jurisdiction['state']).count() > 1:
```
Comment (jcushman) [inline capstone/scripts/ingest_tt_data.py]: Does this mean we'll create a maximum of two jurisdictions with a given state? If so maybe add a comment to explain the goal

**Review by jcushman — COMMENTED** [review]
Some random nitpicks ...

File: capstone/capdb/models.py (hunk 4 above)
Comment (ChefAndy) [reply capstone/capdb/models.py]: I need to change this to a real citation anyway

File: capstone/scripts/ingest_tt_data.py [hunk 8]
```diff
@@ -91,5 +91,21 @@ def copyModel(source, destination, field_map, dupcheck, dupe_field='id'):
...
+
         except IntegrityError as e:
             print(e)
+
+def populate_jurisdiction():
+    """This populates the jurisdiction table based on what's in the tracking tool stub"""
+    reporters = Reporters.objects.values('state').distinct() 
+    for jurisdiction in reporters:
+        # ensures no dupes if the command was already run. Small enough dataset to check every time
+        if Jurisdiction.objects.filter(name=jurisdiction['state']).count() > 1:
```
Comment (anastasia) [inline capstone/scripts/ingest_tt_data.py]: should this be > 0 or >= 1?

Comment (ChefAndy) [reply capstone/scripts/ingest_tt_data.py]: updated

**Pull request merged**